File: cli/internal/connectapi/bots.go
package connectapi

import (
	"context"
	"errors"
	"strings"

	"connectrpc.com/connect"
	"google.golang.org/protobuf/types/known/timestamppb"
	"hmans.de/chatto/internal/core"
	apiv1 "hmans.de/chatto/internal/pb/chatto/api/v1"
)

type botService struct{ api *API }

func (s *botService) ListBots(ctx context.Context, req *connect.Request[apiv1.ListBotsRequest]) (*connect.Response[apiv1.ListBotsResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	bots, err := s.api.core.ListBots(ctx, caller.UserID)
	if err != nil {
		return nil, connectError(err)
	}
	search := strings.ToLower(strings.TrimSpace(req.Msg.GetSearch()))
	if search != "" {
		filtered := bots[:0]
		for _, bot := range bots {
			if strings.Contains(strings.ToLower(bot.User.GetLogin()), search) || strings.Contains(strings.ToLower(bot.User.GetDisplayName()), search) {
				filtered = append(filtered, bot)
			}
		}
		bots = filtered
	}
	limit, offset := apiPagination(req.Msg.GetPage(), 20, 100)
	page, total, more := apiSlicePage(bots, limit, offset)
	result, err := newBotAssembler(s.api).assemble(ctx, page)
	if err != nil {
		return nil, err
	}
	return connect.NewResponse(&apiv1.ListBotsResponse{Bots: result, Page: apiPageInfo(total, more)}), nil
}

func (s *botService) GetBot(ctx context.Context, req *connect.Request[apiv1.GetBotRequest]) (*connect.Response[apiv1.GetBotResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	bot, err := s.api.core.GetBot(ctx, caller.UserID, req.Msg.GetBotUserId())
	if err != nil {
		return nil, connectError(err)
	}
	mapped, err := newBotAssembler(s.api).assembleOne(ctx, bot)
	if err != nil {
		return nil, err
	}
	return connect.NewResponse(&apiv1.GetBotResponse{Bot: mapped}), nil
}

func (s *botService) BatchGetBots(ctx context.Context, req *connect.Request[apiv1.BatchGetBotsRequest]) (*connect.Response[apiv1.BatchGetBotsResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	seen := make(map[string]struct{})
	bots := make([]*core.Bot, 0, len(req.Msg.GetBotUserIds()))
	for _, id := range req.Msg.GetBotUserIds() {
		if _, ok := seen[id]; ok {
			continue
		}
		seen[id] = struct{}{}
		bot, err := s.api.core.GetBot(ctx, caller.UserID, id)
		if err != nil {
			if errors.Is(err, core.ErrNotFound) || errors.Is(err, core.ErrPermissionDenied) {
				continue
			}
			return nil, connectError(err)
		}
		bots = append(bots, bot)
	}
	result, err := newBotAssembler(s.api).assemble(ctx, bots)
	if err != nil {
		return nil, err
	}
	return connect.NewResponse(&apiv1.BatchGetBotsResponse{Bots: result}), nil
}

func (s *botService) CreateBot(ctx context.Context, req *connect.Request[apiv1.CreateBotRequest]) (*connect.Response[apiv1.CreateBotResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	bot, err := s.api.core.CreateBotWithAPIKeyName(ctx, caller.UserID, req.Msg.GetLogin(), req.Msg.GetDisplayName(), req.Msg.GetApiKeyName())
	if err != nil {
		return nil, connectError(err)
	}
	mapped, err := apiBot(ctx, s.api, bot)
	if err != nil {
		return nil, err
	}
	return connect.NewResponse(&apiv1.CreateBotResponse{
		Bot: mapped, ApiKey: bot.APIKey, ApiKeyMetadata: apiBotAPIKeyByID(mapped, bot.APIKeyID),
	}), nil
}

func (s *botService) DeleteBot(ctx context.Context, req *connect.Request[apiv1.DeleteBotRequest]) (*connect.Response[apiv1.DeleteBotResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	deleted, err := s.api.core.DeleteBot(ctx, caller.UserID, req.Msg.GetBotUserId())
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&apiv1.DeleteBotResponse{Deleted: deleted}), nil
}

func (s *botService) CreateBotApiKey(ctx context.Context, req *connect.Request[apiv1.CreateBotApiKeyRequest]) (*connect.Response[apiv1.CreateBotApiKeyResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	issued, err := s.api.core.CreateBotAPIKey(ctx, caller.UserID, req.Msg.GetBotUserId(), req.Msg.GetName())
	if err != nil {
		return nil, connectError(err)
	}
	mapped, err := apiBot(ctx, s.api, issued.Bot)
	if err != nil {
		return nil, err
	}
	return connect.NewResponse(&apiv1.CreateBotApiKeyResponse{
		Bot: mapped, ApiKey: issued.Credential, ApiKeyMetadata: apiBotAPIKeyByID(mapped, issued.KeyID),
	}), nil
}

func (s *botService) RevokeBotApiKey(ctx context.Context, req *connect.Request[apiv1.RevokeBotApiKeyRequest]) (*connect.Response[apiv1.RevokeBotApiKeyResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	bot, err := s.api.core.RevokeBotAPIKey(ctx, caller.UserID, req.Msg.GetBotUserId(), req.Msg.GetKeyId())
	if err != nil {
		return nil, connectError(err)
	}
	mapped, err := newBotAssembler(s.api).assembleOne(ctx, bot)
	if err != nil {
		return nil, err
	}
	return connect.NewResponse(&apiv1.RevokeBotApiKeyResponse{Bot: mapped}), nil
}

func (s *botService) CreateBotIncomingWebhook(ctx context.Context, req *connect.Request[apiv1.CreateBotIncomingWebhookRequest]) (*connect.Response[apiv1.CreateBotIncomingWebhookResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	issued, err := s.api.core.CreateBotIncomingWebhook(ctx, caller.UserID, req.Msg.GetBotUserId(), req.Msg.GetName())
	if err != nil {
		return nil, connectError(err)
	}
	mapped, err := apiBot(ctx, s.api, issued.Bot)
	if err != nil {
		return nil, err
	}
	return connect.NewResponse(&apiv1.CreateBotIncomingWebhookResponse{
		Bot: mapped, WebhookUrl: s.incomingWebhookURL(ctx, issued.Credential),
	}), nil
}

func (s *botService) RevokeBotIncomingWebhook(ctx context.Context, req *connect.Request[apiv1.RevokeBotIncomingWebhookRequest]) (*connect.Response[apiv1.RevokeBotIncomingWebhookResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	bot, err := s.api.core.RevokeBotIncomingWebhook(ctx, caller.UserID, req.Msg.GetBotUserId(), req.Msg.GetWebhookId())
	if err != nil {
		return nil, connectError(err)
	}
	mapped, err := newBotAssembler(s.api).assembleOne(ctx, bot)
	if err != nil {
		return nil, err
	}
	return connect.NewResponse(&apiv1.RevokeBotIncomingWebhookResponse{Bot: mapped}), nil
}

func (s *botService) incomingWebhookURL(ctx context.Context, credential string) string {
	return s.api.absolutizeServerURL(ctx, "/webhooks/incoming/"+credential)
}

func (s *botService) ReassignBotOwner(ctx context.Context, req *connect.Request[apiv1.ReassignBotOwnerRequest]) (*connect.Response[apiv1.ReassignBotOwnerResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	bot, err := s.api.core.ReassignBotOwner(ctx, caller.UserID, req.Msg.GetBotUserId(), req.Msg.GetOwnerUserId())
	if err != nil {
		return nil, connectError(err)
	}
	mapped, err := newBotAssembler(s.api).assembleOne(ctx, bot)
	if err != nil {
		return nil, err
	}
	return connect.NewResponse(&apiv1.ReassignBotOwnerResponse{Bot: mapped}), nil
}

func apiBot(ctx context.Context, api *API, bot *core.Bot) (*apiv1.Bot, error) {
	user, err := requiredUserSummary(ctx, api, bot.User)
	if err != nil {
		return nil, err
	}
	out := &apiv1.Bot{User: user, OwnerUserId: bot.OwnerUserID, CreatedAt: bot.User.GetCreatedAt(), ApiKeyCreatedAt: timestamppb.New(bot.APIKeyCreatedAt)}
	for _, key := range bot.APIKeys {
		mapped := &apiv1.BotApiKey{Id: key.ID, Name: key.Name, CreatedAt: timestamppb.New(key.CreatedAt)}
		switch key.LastUsedState {
		case core.BotCredentialLastUsedNoUseRecorded:
			mapped.LastUsedState = apiv1.CredentialLastUsedState_CREDENTIAL_LAST_USED_STATE_NO_USE_RECORDED
		case core.BotCredentialLastUsedRecorded:
			mapped.LastUsedState = apiv1.CredentialLastUsedState_CREDENTIAL_LAST_USED_STATE_RECORDED
			mapped.LastUsedAt = timestamppb.New(key.LastUsedAt)
		case core.BotCredentialLastUsedUnavailable:
			mapped.LastUsedState = apiv1.CredentialLastUsedState_CREDENTIAL_LAST_USED_STATE_UNAVAILABLE
		}
		out.ApiKeys = append(out.ApiKeys, mapped)
	}
	for _, webhook := range bot.IncomingWebhooks {
		mapped := &apiv1.BotIncomingWebhook{
			Id: webhook.ID, Name: webhook.Name, CreatedAt: timestamppb.New(webhook.CreatedAt),
		}
		switch webhook.LastUsedState {
		case core.BotCredentialLastUsedNoUseRecorded:
			mapped.LastUsedState = apiv1.CredentialLastUsedState_CREDENTIAL_LAST_USED_STATE_NO_USE_RECORDED
		case core.BotCredentialLastUsedRecorded:
			mapped.LastUsedState = apiv1.CredentialLastUsedState_CREDENTIAL_LAST_USED_STATE_RECORDED
			mapped.LastUsedAt = timestamppb.New(webhook.LastUsedAt)
		case core.BotCredentialLastUsedUnavailable:
			mapped.LastUsedState = apiv1.CredentialLastUsedState_CREDENTIAL_LAST_USED_STATE_UNAVAILABLE
		}
		out.IncomingWebhooks = append(out.IncomingWebhooks, mapped)
	}
	return out, nil
}

func apiBotAPIKeyByID(bot *apiv1.Bot, keyID string) *apiv1.BotApiKey {
	if bot == nil {
		return nil
	}
	for _, key := range bot.GetApiKeys() {
		if key.GetId() == keyID {
			return key
		}
	}
	return nil
}

File: apps/frontend/src/lib/api-client/bots.ts
import { authHeaders, createChattoClient } from './connect.js';
import { BotService } from '@chatto/api-types/api/v1/bots_connect';
import { CredentialLastUsedState, type Bot as APIBot } from '@chatto/api-types/api/v1/bots_pb';

export type BotAPIConfig = {
  baseUrl: string;
  bearerToken: string | null;
  onAuthenticationRequired?: (serverId: string) => void;
};

export type Bot = {
  id: string;
  login: string;
  displayName: string;
  avatarUrl: string | null;
  bio: string | null;
  timezone: string | null;
  ownerUserId: string;
  createdAt: Date | null;
  apiKeyCreatedAt: Date | null;
  apiKeys: {
    id: string;
    name: string;
    createdAt: Date | null;
    lastUsedState: 'no_use_recorded' | 'recorded' | 'unavailable';
    lastUsedAt: Date | null;
  }[];
  incomingWebhooks: {
    id: string;
    name: string;
    createdAt: Date | null;
    lastUsedState: 'no_use_recorded' | 'recorded' | 'unavailable';
    lastUsedAt: Date | null;
  }[];
};

export type BotPage = {
  bots: Bot[];
  totalCount: number;
  hasMore: boolean;
};

export function createBotAPI(config: BotAPIConfig) {
  const client = createChattoClient(BotService, config);
  const headers = () => authHeaders(config);
  return {
    async listBots(
      input: { search?: string | null; limit: number; offset: number },
      options: { signal?: AbortSignal } = {}
    ): Promise<BotPage> {
      const response = await client.listBots(
        {
          search: input.search ?? '',
          page: { limit: input.limit, offset: input.offset }
        },
        { headers: headers(), ...(options.signal ? { signal: options.signal } : {}) }
      );
      return {
        bots: response.bots.map(botFromAPI),
        totalCount: Number(response.page?.totalCount ?? response.bots.length),
        hasMore: response.page?.hasMore ?? false
      };
    },
    async getBot(botUserId: string, options: { signal?: AbortSignal } = {}): Promise<Bot> {
      const response = await client.getBot(
        { botUserId },
        { headers: headers(), ...(options.signal ? { signal: options.signal } : {}) }
      );
      return botFromAPI(requiredBot(response.bot));
    },
    async createBot(input: {
      login: string;
      displayName: string;
      apiKeyName?: string;
    }): Promise<{ bot: Bot; apiKey: string }> {
      const response = await client.createBot(input, { headers: headers() });
      return { bot: botFromAPI(requiredBot(response.bot)), apiKey: response.apiKey };
    },
    async deleteBot(botUserId: string): Promise<boolean> {
      return (await client.deleteBot({ botUserId }, { headers: headers() })).deleted;
    },
    async createBotAPIKey(botUserId: string, name: string): Promise<{ bot: Bot; apiKey: string }> {
      const response = await client.createBotApiKey({ botUserId, name }, { headers: headers() });
      return { bot: botFromAPI(requiredBot(response.bot)), apiKey: response.apiKey };
    },
    async revokeBotAPIKey(botUserId: string, keyId: string): Promise<Bot> {
      const response = await client.revokeBotApiKey({ botUserId, keyId }, { headers: headers() });
      return botFromAPI(requiredBot(response.bot));
    },
    async createBotIncomingWebhook(
      botUserId: string,
      name: string
    ): Promise<{ bot: Bot; webhookUrl: string }> {
      const response = await client.createBotIncomingWebhook(
        { botUserId, name },
        { headers: headers() }
      );
      return { bot: botFromAPI(requiredBot(response.bot)), webhookUrl: response.webhookUrl };
    },
    async revokeBotIncomingWebhook(botUserId: string, webhookId: string): Promise<Bot> {
      const response = await client.revokeBotIncomingWebhook(
        { botUserId, webhookId },
        { headers: headers() }
      );
      return botFromAPI(requiredBot(response.bot));
    },
    async reassignBotOwner(botUserId: string, ownerUserId: string): Promise<Bot> {
      const response = await client.reassignBotOwner(
        { botUserId, ownerUserId },
        { headers: headers() }
      );
      return botFromAPI(requiredBot(response.bot));
    }
  };
}

export type BotAPI = ReturnType<typeof createBotAPI>;

function requiredBot(bot: APIBot | undefined): APIBot {
  if (!bot?.user) throw new Error('bot response did not include bot metadata');
  return bot;
}

function botFromAPI(bot: APIBot): Bot {
  const user = bot.user;
  if (!user) throw new Error('bot response did not include a user');
  return {
    id: user.id,
    login: user.login,
    displayName: user.displayName,
    avatarUrl: user.avatarUrl ?? null,
    bio: user.bio ?? null,
    timezone: user.timezone ?? null,
    ownerUserId: bot.ownerUserId,
    createdAt: bot.createdAt?.toDate() ?? null,
    apiKeyCreatedAt: bot.apiKeyCreatedAt?.toDate() ?? null,
    apiKeys: (bot.apiKeys ?? []).map((key) => ({
      id: key.id,
      name: key.name,
      createdAt: key.createdAt?.toDate() ?? null,
      lastUsedState:
        key.lastUsedState === CredentialLastUsedState.RECORDED
          ? 'recorded'
          : key.lastUsedState === CredentialLastUsedState.NO_USE_RECORDED
            ? 'no_use_recorded'
            : 'unavailable',
      lastUsedAt: key.lastUsedAt?.toDate() ?? null
    })),
    incomingWebhooks: (bot.incomingWebhooks ?? []).map((webhook) => ({
      id: webhook.id,
      name: webhook.name,
      createdAt: webhook.createdAt?.toDate() ?? null,
      lastUsedState:
        webhook.lastUsedState === CredentialLastUsedState.RECORDED
          ? 'recorded'
          : webhook.lastUsedState === CredentialLastUsedState.NO_USE_RECORDED
            ? 'no_use_recorded'
            : 'unavailable',
      lastUsedAt: webhook.lastUsedAt?.toDate() ?? null
    }))
  };
}
